odo/backends/avro.py

```diff
@@ -0,0 +1,252 @@
+
+        Returns
+        -------
+        avro.datafile.DataFileWriter
+        """
+        rec_writer = io.DatumWriter()
+        df_writer = datafile.DataFileWriter(
+            open(uri, 'ab+'),
+            rec_writer
+        )
+        #Check for embedded schema to ensure existing file is an avro file.
+        embedded_schema = schema.parse(df_writer.get_meta('avro.schema'))
+
+        #If writers_schema supplied, check for equality with embedded schema.
+        if writers_schema:
+            assert embedded_schema == writers_schema, \
+                "writers_schema embedded in {uri} differs from user supplied schema for appending."
+
+        return df_writer
+
+    @staticmethod
+    def _get_new_writer(uri, sch):
+        """
+        Returns an isntance of avro.datafile.DataFileWriter for writing
+        to a new avro file at `uri`.
+
+        Parameters
+        ----------
+
+        uri : str
+            uri of avro existing, non-empty avro file
+
+        sch : avro.schema.Schema object
+
+        Returns
+        -------
+        avro.datafile.DataFileWriter
+        """
+        rec_writer = io.DatumWriter()
+        df_writer = datafile.DataFileWriter(
+            open(uri, 'wb'),
+            rec_writer,
+            writers_schema = sch
+        )
+        return df_writer
+
+    @property
+    def writer(self):
+        if hasattr(self, '_writer'):
+            return self._writer
+        else:
+            if os.path.exists(self.uri) and os.path.getsize(self.uri) > 0:
+                df_writer = self._get_append_writer(self.uri, self.schema)
+            else:
+                df_writer = self._get_new_writer(self.uri, self.schema)
+            self._writer = df_writer
+        return df_writer
+
+    def flush(self):
+        if hasattr(self, '_writer'):
+            self._writer.close()
+            del(self._writer)
+
+
+@resource.register('.+\.(avro)')
+def resource_avro(uri, schema=None, **kwargs):
+    return AVRO(uri, schema=schema, **kwargs)
+
+def discover_schema(sch):
+    if isinstance(sch, schema.RecordSchema):
+        return var * Record([(f.name, discover_schema(f.type)) for f in sch.fields])
+    elif isinstance(sch, schema.UnionSchema):
+        try:
+            types = [s.type for s in sch.schemas]
+            assert "null" in types
+            types.remove("null")
+            assert len(types) == 1
+            return Option(AVRO_TYPE_MAP[types[0]])
+        except AssertionError:
+            raise TypeError("odo supports avro UnionSchema only for nullabel fields.  "
+                            "Received {0}".format(str([s.type for s in sch.schemas])))
+    elif isinstance(sch, schema.PrimitiveSchema):
+        return AVRO_TYPE_MAP[sch.type]
+    elif isinstance(sch, schema.MapSchema):
+        return Map(string, discover_schema(sch.values))
+    elif isinstance(sch, schema.ArraySchema):
+        return var * discover_schema(sch.items)
+    else:
+        raise Exception(str(type(sch)))
```

Comment:
And should probably say something like `'Unable to discover avro type %r' % type(sch).__name__`